storage/disk/txn.go

```diff
@@ -0,0 +1,423 @@
+}
+
+func newTransaction(xid uint64, write bool, underlying *badger.Txn, context *storage.Context, pm *pathMapper, trie *partitionTrie, db *Store) *transaction {
+	return &transaction{
+		underlying: underlying,
+		partitions: trie,
+		pm:         pm,
+		db:         db,
+		xid:        xid,
+		stale:      false,
+		write:      write,
+		context:    context,
+	}
+}
+
+func (txn *transaction) ID() uint64 {
+	return txn.xid
+}
+
+func (txn *transaction) Commit(_ context.Context) (storage.TriggerEvent, error) {
+	// TODO(tsandall): This implementation does not provide any data or policy
+	// events because they are of minimal value given that the transcation
+	// cannot be used for reads once the commit finishes. This differs from the
+	// in-memory store.
+	//
+	// We should figure out how to remove the code in the plugin manager that
+	// performs reads on committed transactions.
+	txn.stale = true
+	return storage.TriggerEvent{Context: txn.context}, wrapError(txn.underlying.Commit())
+}
+
+func (txn *transaction) Abort(_ context.Context) {
+	txn.stale = true
+	txn.underlying.Discard()
+}
+
+func (txn *transaction) Read(_ context.Context, path storage.Path) (interface{}, error) {
+
+	i, node := txn.partitions.Find(path)
+
+	if node == nil {
+
+		key, err := txn.pm.DataPath2Key(path[:i])
+		if err != nil {
+			return nil, err
+		}
+
+		value, err := txn.readOne(key)
+		if err != nil {
+			return nil, err
+		}
+
+		return ptr.Ptr(value, path[i:])
+	}
+
+	key, err := txn.pm.DataPrefix2Key(path[:i])
+	if err != nil {
+		return nil, err
+	}
+
+	return txn.readMultiple(i, key)
+}
+
+func (txn *transaction) readMultiple(offset int, prefix []byte) (interface{}, error) {
+
+	result := map[string]interface{}{}
+
+	it := txn.underlying.NewIterator(badger.IteratorOptions{Prefix: prefix})
+	defer it.Close()
+
+	var keybuf, valbuf []byte
+
+	for it.Rewind(); it.Valid(); it.Next() {
+
+		keybuf = it.Item().KeyCopy(keybuf)
+		path, err := txn.pm.DataKey2Path(keybuf)
+		if err != nil {
+			return nil, err
+		}
+
+		valbuf, err = it.Item().ValueCopy(valbuf)
+		if err != nil {
+			return nil, wrapError(err)
+		}
+
+		var value interface{}
+		if err := deserialize(valbuf, &value); err != nil {
+			return nil, err
+		}
+
+		node := result
+
+		for i := offset; i < len(path)-1; i++ {
+			child, ok := node[path[i]]
+			if !ok {
+				child = map[string]interface{}{}
+				node[path[i]] = child
+			}
+			childObj, ok := child.(map[string]interface{})
+			if !ok {
+				return nil, &storage.Error{Code: storage.InternalErr, Message: fmt.Sprintf("corrupt key-value: %s", keybuf)}
+			}
+			node = childObj
+		}
+
+		node[path[len(path)-1]] = value
+	}
+
+	if len(result) == 0 {
+		return nil, errNotFound
+	}
+
+	return result, nil
+}
+
+func (txn *transaction) readOne(key []byte) (interface{}, error) {
+
+	item, err := txn.underlying.Get(key)
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			return nil, errNotFound
+		}
+		return nil, wrapError(err)
```

Comment:
💭 Looks like a thing `wrapError` could take care of: rewrapping badger.ErrKeyNotFound into our errNotFound.

Reply:
I thought about that and opted to _not_ add that behaviour in wrapError because it felt like an important implementation detail and should not be hidden behind the relatively generic wrapError function which is a catch-all... in other words, I'm not sure that all `badger.ErrKeyNotFound` errors should be treated as `storage.NotFoundError` (or whatever it's called).

Reply:
I'll amend with a comment in wrapError to capture this.